### DISCUSSION TITLE
Neat Alias: Importing a markdown file as a note

### DISCUSSION THREAD

**Discussion by nogweii** [zk-org/zk#97] [Share]
Here is a neat alias I've been using to some success for bringing random markdown files across my system into my notebook, using 0.7.0:

```
[alias]
# Import a markdown file into the notebook, generating an ID for the fresh note
import-md = "cat $OLDPWD/$1 | zk new --title \"$(.zk/bin/markdown-title.rb $OLDPWD/$1)\" $2 && rm $OLDPWD/$1"
```

This relies on a Ruby script I wrote that retrieves the title of a markdown file, either by fetching it from the YAML front matter or by using the first header:

```ruby
#!/usr/bin/env ruby

require 'date'
require 'kramdown'

doc_title = nil
file_content = File.open(ARGV[0]).read

YAML_BLOCK = %r!\A(---\s*\n.*?\n?)^((---|\.\.\.)\s*$\n?)!m.freeze
yaml_match = file_content.match(YAML_BLOCK)

if yaml_match
  yaml_frontmatter = YAML.safe_load(yaml_match[1], permitted_classes: [Date, Time])

  front_matter_title = yaml_frontmatter["title"]
  unless front_matter_title.nil? or front_matter_title.empty?
    doc_title = front_matter_title
  end
  markdown_content = yaml_match.post_match
else
  markdown_content = file_content
end

unless doc_title
  doc = Kramdown::Document.new(markdown_content)
  toc = doc.to_toc
  doc_title = toc.children.first.value.options[:raw_text]
end

puts doc_title
```

With this, I can run `zk import-md foo.md` and have it moved to my `$ZK_NOTEBOOK_DIR` and open in Neovim. Very nice.

**Comment by nogweii** [comment]
@mickael-menu one thing I realized in building this is that `zk` will cd to the notebook directory, so I have to use a trick of zsh (`$OLDPWD`) to get the absolute path to the file I specified in the CLI. It would be helpful if zk would preserve the prior directory it was in before changing directories.

e.g. `zk` exports an additional variable, `ZK_RUNNING_DIRECTORY` or some other name, that stores the name of where the user ran zk from, not where the notebook directory is.

**Reply by mickael-menu** [reply]
Thanks for sharing this alias!

> one thing I realized in building this is that `zk` will cd to the notebook directory, so I have to use a trick of zsh (`$OLDPWD`) to get the absolute path to the file I specified in the CLI.

I chose to `cd` to the notebook directory explicitly to be able to write the alias shell code as if it was running from the root of the notebook. This makes sure it works no matter where you call `zk`.

https://github.com/mickael-menu/zk/blob/59b8269344363dca2439cb168e847aa9f35f4444/main.go#L147-L150

I didn't think about the case where you would want to use a file from the current directory as argument though... We could definitely save the previous directory in a variable as you suggested. But I'm wondering what should be the default behavior. Intuitively you expected the alias to run in the context of the shell wd or the notebook dir?

**Reply by nogweii** [reply]
Either default makes sense to me, and making it change to the notebook directory feels a little more convenient. Also, it's the backwards-compatible one.

**Reply by mickael-menu** [reply]
Yeah, backward-compatibility is definitely a plus there, although I allow myself breaking changes until we reach the 1.0.
